src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -268,65 +265,125 @@ class Checker(globalTable: SymbolTable, verbose: Boolean = false) {
-    private def contractIsSubtype(c1: String, c2: String): Boolean = {
-        val c1Table = globalTable.contractLookup(c1)
-        val c2Table = globalTable.contractLookup(c2)
+    private def contractIsSubtype(table: ContractTable, c1: ContractType, c2: ContractType, isThis: Boolean): Either[Boolean, Error] = {
+        // Everything is a subtype of Top
+        if (c2.contractName == "Top") {
+            return Left(true)
+        }
 
-        (c1Table.contract, c2Table.contract) match {
-            case (obs1: ObsidianContractImpl, obs2: ObsidianContractImpl) => (obs1 == obs2)
-            case (jvcon1: JavaFFIContractImpl, jvcon2: JavaFFIContractImpl) => (jvcon1 == jvcon2)
-            case (obs: ObsidianContractImpl, jvcon: JavaFFIContractImpl) => false
-            case (jvcon: JavaFFIContractImpl, obs:ObsidianContractImpl) => (obs.name == jvcon.interface)
+        val c1Lookup = globalTable.contract(c1.contractName).map(_.contract)
+        val c2Lookup = globalTable.contract(c2.contractName).map(_.contract)
+
+        if (c1Lookup.isEmpty) {
+            return Right(ContractUndefinedError(c1.contractName))
         }
+
+        if (c1Lookup.isEmpty) {
+            return Right(ContractUndefinedError(c2.contractName))
+        }
+
+        Left((c1Lookup.get, c2Lookup.get) match {
+            case (obs1: ObsidianContractImpl, obs2: ObsidianContractImpl) =>
+                if (!obs1.isInterface && obs2.isInterface) {
+                    val interfaceMatches = obs1.bound.contractName == obs2.name
+
+                    val subsParams = obs1.bound.substitute(obs1.params, c1.typeArgs).typeArgs
+                    val argChecks = subsParams.zip(c2.typeArgs).forall(checkArgCompat)
+
+                    interfaceMatches && argChecks
+                } else {
+                   obs1 == obs2 && c1.typeArgs.zip(c2.typeArgs).forall(checkArgCompat)
+                }
+            case (jvcon1: JavaFFIContractImpl, jvcon2: JavaFFIContractImpl) => jvcon1 == jvcon2
+            case (obs: ObsidianContractImpl, jvcon: JavaFFIContractImpl) => false
+            case (jvcon: JavaFFIContractImpl, obs:ObsidianContractImpl) => obs.name == jvcon.interface
+        })
     }
 
+    private def typeBound(table: ContractTable): ObsidianType => ObsidianType = {
+        case np: NonPrimitiveType => np match {
+            case GenericType(gVar, bound) => bound.referenceType
+            case t => t
+        }
+
+        case t => t
+    }
 
     /* true iff [t1 <: t2] */
-    private def isSubtype(t1: ObsidianType, t2: ObsidianType, isThis: Boolean): Option[Error] = {
-        (t1, t2) match {
-            case (BottomType(), _) => None
-            case (_, BottomType()) => None
-            case (IntType(), IntType()) => None
-            case (BoolType(), BoolType()) => None
-            case (StringType(), StringType()) => None
+    private def isSubtype(table: ContractTable, t1: ObsidianType, t2: ObsidianType, isThis: Boolean): Option[Error] = {
```

Comment:
As a separate PR, if you were in the mood, you could refactor this. A method called isSubtype should return a Bool, not an Option[Error].